### ISSUE TITLE
RKE2 - AWS Cloud Provider not working - clusters stuck in provisioning state

### ISSUE THREAD

**Issue by davidnuzik** [rancher/rancher#34105]
**Environment:**
Rancher version: v2.6-head 4b6af95d7 8/9/21 9:25 am Pacific
Rancher cluster type: single-node docker
Docker version: 19.03
 
Downstream cluster type: rke2
Downstream K8s version: v1.21.3-rc4+rke2r2
 
**Steps to reproduce:**
1. Create an AWS RKE2 cluster in Rancher a 1-node cluster is fine. You will be prompted to create cloud credentials if you haven't yet. In my case I used region us-east-2 in the QA aws account. Specify aws as the cloud provider in the UI. Be sure to provide a valid IAM role as well as this is required for the aws cloud provider. I used instance type t3.xlarge.

![image](https://user-images.githubusercontent.com/45179589/128791337-7174ecc5-4164-4260-9cb8-c07353766a6b.png)

2. The cluster will be stuck in the Provisioning state forever. The RKE2 config seems to look good to me, the aws cloud provider was set on the node.

![image](https://user-images.githubusercontent.com/45179589/128791386-de699c0c-edd5-492d-96cb-3ec82837af91.png)

**Additional Info:**
If you do the same with an RKE1 cluster it will provision correctly, become active, and I can indeed use cloud provider features.

When testing, I didn't observe anything particularly helpful in the Rancher Logs. I can provide logs upon request.
When testing, the UI appeared to send correct request payload when creating the RKE2 cluster. I saw the fleet web request go in and this specified the IAM profile correctly and such. Then the cluster creation request and it appeared correct as well. I can provide the content upon request.

**Comment by davidnuzik** [comment]
Never-mind, only IAM profile has to be given.

**Comment by galal-hussein** [comment]
The issue comes down to this:

RKE2 kubelet tries to register itself with a node name different than the one configured in the cloud provider, and here are my findings:

## In RKE1

- We use only one certificate for all kubelets for all nodes with the OU,CN like that:
```
Subject: O = system:nodes, CN = system:node
```
Which means kubelet wont be authorized by the node authorizer as it needs the CN to be in this format CN = system:node:<node-name>

- Also RKE1 adds a specific cluster role binding to allow RBAC to authorize all kubelets: https://github.com/rancher/rke/blob/master/templates/authz.go#L32-L47

Thats how kubelet can register itself on aws cloud provider in rke1

## In RKE2

- each node has different certificate for kubelet with following ou,cn:
```
Subject: O = system:nodes, CN = system:node:ip-172-31-15-189
```
- RKE2 uses `hostname` as the hostname-override as a kubelet argument internally

- Since these OU and CN are set by this format, Node Restriction admission plugin kicks in and prevent registration of the node:

For more information about the issue:

https://kubernetes.io/docs/reference/access-authn-authz/node/
https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction

## Workaround suggestion

The workaround can be as following:

- If aws cloud provider is used in rke2
- if user didnt add --node-name

Then RKE2 can use code equivalent to `hostname -f` instead of `hostname` to set for the node name.

**Comment by Jono-SUSE-Rancher** [comment]
Ahh thanks Craig, didn't see this. I think the process would have us move this into the waiting for RKE2 bucket. Even though @kinarashah has some work to do here, we will need to track when the kubelet error gets fixed.

**Comment by kinarashah** [comment]
Available to test with `v1.21.3-rc6+rke2r2`.

**Comment by davidnuzik** [comment]
**My checks PASSED**
 
**Reproduction Steps:**

Not required.
 
**Validation Steps:**

Rancher version: v2.6-head `72986b2e4` pulled at 8/17/21 1:32 pm Pacific
Rancher cluster type: single-node docker
Docker version: 20.10
 
Downstream cluster type: rke2
Downstream K8s version: v1.21.3-rc6+rke2r2
 
Validation steps:

1. Create a 1-node rke2 cluster with the aws node driver. Specify the "Amazon" cloud provider and provide a suitable IAM role.
2. Note, the cluster is no longer stuck in the Provisioning state (the bug is fixed).
3. Note, I can successfully leverage cloud provider features, such as setting up a storage class and pvc and see my storage object created in aws web console.
![image](https://user-images.githubusercontent.com/45179589/129804829-efc5ca8b-ee96-4c51-bae7-585da5e8cfe4.png)
![image](https://user-images.githubusercontent.com/45179589/129804872-d5454bd1-59b0-42f4-8168-e813b916213b.png)


Additional Info:
During testing, ensured UI is passing correct request payloads. Nothing looked abnormal on the UI side during my testing.

Further testing of the aws cloud provider on RKE2 will be done via https://github.com/rancher/rancher/issues/32390
